enterprise/cmd/frontend/internal/batches/httpapi/file_handler.go

```diff
@@ -0,0 +1,235 @@
+package httpapi
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/opentracing/opentracing-go/log"
+	sglog "github.com/sourcegraph/log"
+
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/batches/store"
+	btypes "github.com/sourcegraph/sourcegraph/enterprise/internal/batches/types"
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/internal/errcode"
+	"github.com/sourcegraph/sourcegraph/internal/observation"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// FileHandler handles retrieving and uploading of files.
+type FileHandler struct {
+	logger     sglog.Logger
+	db         database.DB
+	store      BatchesStore
+	operations *Operations
+}
+
+type BatchesStore interface {
+	CountBatchSpecWorkspaceFiles(context.Context, store.ListBatchSpecWorkspaceFileOpts) (int, error)
+	GetBatchSpec(context.Context, store.GetBatchSpecOpts) (*btypes.BatchSpec, error)
+	GetBatchSpecWorkspaceFile(context.Context, store.GetBatchSpecWorkspaceFileOpts) (*btypes.BatchSpecWorkspaceFile, error)
+	UpsertBatchSpecWorkspaceFile(context.Context, *btypes.BatchSpecWorkspaceFile) error
+}
+
+// NewFileHandler creates a new FileHandler.
+func NewFileHandler(db database.DB, store BatchesStore, operations *Operations) http.Handler {
+	handler := &FileHandler{
+		logger:     sglog.Scoped("FileHandler", ""),
+		db:         db,
+		store:      store,
+		operations: operations,
+	}
+	return handler
+}
+
+const maxUploadSize = 10 << 20 // 10MB
+
+func (h *FileHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Wrap the method operations to get a single point for logging.
+	responseBody, statusCode, err := func() (_ io.Reader, statusCode int, err error) {
+		ctx, _, endObservation := h.operations.serveHTTP.With(r.Context(), &err, observation.Args{})
+		defer func() {
+			endObservation(1, observation.Args{LogFields: []log.Field{
+				log.String("method", r.Method),
+				log.Int("statusCode", statusCode),
+			}})
+		}()
+
+		var handlerFunc fileHandlerFunc
+
+		switch r.Method {
+		case http.MethodGet:
+			handlerFunc = h.get
+		case http.MethodHead:
+			handlerFunc = h.exists
+		case http.MethodPost:
+			// Prevent client from uploading files that are too large. This is also enforced on the src-cli side as well.
+			r.Body = http.MaxBytesReader(w, r.Body, maxUploadSize)
+			//h.upload(w, r)
+			handlerFunc = h.upload
+		default:
+			return nil, http.StatusMethodNotAllowed, nil
+		}
+		return handlerFunc(ctx, r)
+	}()
+
+	if err != nil {
+		http.Error(w, err.Error(), statusCode)
+		return
+	}
+
+	w.WriteHeader(statusCode)
+
+	if responseBody != nil {
+		if _, err := io.Copy(w, responseBody); err != nil {
+			h.logger.Error("batches.httpapi: failed to write payload to client", sglog.Error(err))
```

Comment:
I don't think we need the log message prefix here, given the use of a scoped logger:

```suggestion
			h.logger.Error("failed to write payload to client", sglog.Error(err))
```